Intelligent Az Interactive with Recommendation ability (main)

wu5bocheng opened the pull request:
Az interactive next --fix small bugs and commit one suggestion at a time
[Add] auto complete param value in scenario

wu5bocheng committed:
Merge branch 'interactive-next' of https://github.com/ReaNAiveD/azure-cli-extensions into interactive-next

wu5bocheng committed:
[Fix] fix CryptographyDeprecationWarning by ignoring it.

wu5bocheng committed:
[Feature] recommend one param for the scenario at a time

wu5bocheng committed:
[Feature] give notice to users that they can skip a command or quit the scenario

wu5bocheng committed:
[Update] update notice to users that they can skip a command or quit the scenario

wu5bocheng committed:
[Update] update notice to users that they can skip a command or quit the scenario

wu5bocheng committed:
[Update] crypto already fixed

wu5bocheng committed:
[Add] Load command table before display the shell

wu5bocheng committed:
Merge branch 'interactive-main' into interactive-next

wu5bocheng committed:
Update scenario_suggest.py

wu5bocheng committed:
[Add] auto complete param value in scenario

wu5bocheng committed:
[Add] auto complete param value in scenario

wu5bocheng committed:
[Fix] fix double print of notice for recommendation

wu5bocheng committed:
[Add] Optimize loading table using processing bar

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -151,8 +155,11 @@ def __call__(self):
-            print("\n\nAny comments or concerns? You can use the \'feedback\' command!" +
+            print("\nAny comments or concerns? You can use the \'feedback\' command!" +
                   " We would greatly appreciate it.\n")
+        if self.cli_ctx.config.getboolean("interactive", "enable_recommender", fallback=True) == True:
```
```suggestion
        if self.cli_ctx.config.getboolean("interactive", "enable_recommender", fallback=True):
```

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -151,8 +155,11 @@ def __call__(self):
+            print("\nAny comments or concerns? You can use the \'feedback\' command!" +
                   " We would greatly appreciate it.\n")
+        if self.cli_ctx.config.getboolean("interactive", "enable_recommender", fallback=True) == True:
+            print("\nA new Recommender is added which can make the completion ability more intelligent and provide the scenario completion!\n"
+                  "If you don't want to enable this feature, you can use 'az config set interactive.enable_recommender=False' to disable it.\n")
```
```suggestion
            print("\nA new Recommender is added which can make the completion ability more intelligent and provide the scenario completion!\n"
                  "If you want to disable this feature, you can use 'az config set interactive.enable_recommender=False' to disable it.\n")
```

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -815,8 +826,21 @@ def run(self):
+        print_styled_text([(Style.ACTION, "Loading command table... Expected time is around 1 minute.")])
+        time_spent_on_loading = 0
+        # still loading commands, show the time of loading
+        while self.command_table_thread.is_alive() and time_spent_on_loading < 150:
+            time_spent_on_loading = time.time() - self.command_table_thread.start_time
+            # print out loading time every 10 seconds
+            if time_spent_on_loading % 10 < 1:
+                print_styled_text([(Style.ACTION, "Already spent {} seconds on loading command table.".format(round(time_spent_on_loading, 1)))])
+            time.sleep(1)
```
Why not using `time.sleep(10)`
```suggestion
        while self.command_table_thread.is_alive() and time_spent_on_loading < 150:
            time_spent_on_loading = time.time() - self.command_table_thread.start_time
            print_styled_text([(Style.ACTION, "Already spent {} seconds on loading command table.".format(round(time_spent_on_loading, 1)))])
            time.sleep(10)
```

wu5bocheng replied:
Already modified in https://github.com/wu5bocheng/azure-cli-extensions/pull/3
Please review the branch first.

Because I didn't setup new branches for different functions initially. The first few PRS might be confusing.

wu5bocheng committed:
[Add] Optimize loading table using processing bar

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Merge remote-tracking branch 'origin/interactive-next' into interactive-next

wu5bocheng committed:
[Add] Optimize loading table using processing bar

wu5bocheng committed:
[Add] Optimize loading table using processing bar
[Update] Updating value recommendation of global params

wu5bocheng committed:
Merge pull request #3 from wu5bocheng/interactive-autocomplete

Interactive autocomplete

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -78,15 +117,24 @@ def try_remove(container, item):
-                            suggest.append(self.param_value_map[param])
-                    return Suggestion(' '.join(suggest))
+                        if param in self.special_global_param_map.keys():
+                            raw_value = param
```
```suggestion
                            raw_value = self.special_global_param_map[param]
```

wu5bocheng committed:
[Update] fix loading logic to Xing recommend

wu5bocheng committed:
Merge remote-tracking branch 'origin/interactive-next' into interactive-next

wu5bocheng commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -32,12 +40,44 @@ def update(self, sample: str):
+                    cur_param = ''
+                else:
+                    cur_param = self.param_value_map[part]
+                    self.customized_param_value_map[cur_param] = ''
```
self.customized_param_value_map[app] = ' '

wu5bocheng committed:
[Update] Use dictionaries instead of strings to store feedback information

wu5bocheng committed:
[Update] Use dictionaries instead of strings to store feedback information

wu5bocheng committed:
Revert "[Update] Use dictionaries instead of strings to store feedback information"

This reverts commit ffd4b6e7cf3032c69752bbf0c5dfa593110bb630.

wu5bocheng committed:
Revert "[Update] Use dictionaries instead of strings to store feedback information"

This reverts commit 46c742bb4f6f41219728ae170a178fc7ad7d4baa.

wu5bocheng committed:
[Update] Optimize some variable names(rec-->recommendation)
[Add] add api_version to the feedback
[Add] Parsing api_version from api responses
[Update] Optimize feedback function by using dictionary data
[Add] add new property "recommendation" under telemetry "Properties"

wu5bocheng committed:
[Update] Optimize feedback function

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -32,12 +40,44 @@ def update(self, sample: str):
+                        self.customized_param_value_map[param] = ''
+                    cur_param = ''
+                else:
+                    cur_param = self.param_value_map[part]
+                    self.customized_param_value_map[cur_param] = ''
+                    special_global_param_list = None
+            elif cur_param:
+                self.customized_param_value_map[cur_param] += ' ' + part
```
Could you please add some comments to explain the meaning of `param_value_map` and `customized_param_value_map`? Such as, give some examples for their key and value respectively

wu5bocheng replied:
Already modified with the new commit by adding some notes: https://github.com/wu5bocheng/azure-cli-extensions/pull/2/commits/0169eb35aef4ebc0980a79ec1a258244f1deeb8e
and optimizing param name: https://github.com/wu5bocheng/azure-cli-extensions/pull/2/commits/1c9221e24245cdcd46e102d84feba9b1e3a638da

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -78,15 +118,25 @@ def try_remove(container, item):
+                                value = ''
+                            suggests.append({'param': param, 'value': value})
+                    if suggests:
+                        # suggest one parameter at a time
+                        return Suggestion(' '.join([suggests[0]['param'], suggests[0]['value']]))
```
Why don't we just make `suggests` as `{}` type instead of `[]` type?

wu5bocheng replied:
It is because suggests contains a list of recommendations ranging from comands to scenarios, do you mean to use the option_index such as “a1” to the `key` of dictionary `suggests` and suggested data to be `value` of the dict?

zhoxing-ms replied:
We no longer need to define `suggests` as a list, right?

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -78,15 +118,25 @@ def try_remove(container, item):
-                    return Suggestion(' '.join(suggest))
+                        if param in self.special_global_param_map.keys():
+                            # sample_value is the sample values in scenarios, such as <RESOURCEGROUPNAME>
+                            sample_value = param
+                        else:
+                            sample_value = self.param_value_map[param]
+                        if sample_value:
+                            if sample_value in self.customized_param_value_map.keys():
+                                value = self.customized_param_value_map[sample_value]
```
`sample_value` should only be one of the two cases, so it may not be accurate enough. May be the name `cached_param`  more appropriate?

wu5bocheng replied:
Already modified with the new commit by optimizing param name: https://github.com/wu5bocheng/azure-cli-extensions/commit/1c9221e24245cdcd46e102d84feba9b1e3a638da combined with the review comments above.

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -78,15 +118,25 @@ def try_remove(container, item):
-                            suggest.append(self.param_value_map[param])
-                    return Suggestion(' '.join(suggest))
+                        if param in self.special_global_param_map.keys():
+                            # sample_value is the sample values in scenarios, such as <RESOURCEGROUPNAME>
+                            sample_value = param
+                        else:
+                            sample_value = self.param_value_map[param]
```
```suggestion
                            sample_value = param
                        else:
                            # sample_value is the sample values in scenarios, such as <RESOURCEGROUPNAME>
                            sample_value = self.param_value_map[param]
```

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/scenario_suggest.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
[Update] Optimize param value and add notes

from customized_param_value_map to customized_sample_value_map

wu5bocheng committed:
[Update] optimize param value and add some notes

customized_sample_value_map --> customized_cached_param_map

wu5bocheng commented:
Already modified, ready for rereview.

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/recommendation.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
[Update] Optimize feedback property name to parallel with doc

wu5bocheng committed:
[Update] Optimize feedback property name to parallel with doc

wu5bocheng committed:
[Update] Optimize feedback initialize

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -811,12 +823,48 @@ def progress_patch(self, *args, **kwargs):
+                continue_loading = get_yes_or_no_option(step_msg)
+                already_prompted = True
+                # if the customer chooses not to continue loading, break the loading loop
+                if time_loading_too_long and continue_loading == False:
```
```suggestion
                if continue_loading == False:
```

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -811,12 +823,48 @@ def progress_patch(self, *args, **kwargs):
+        # initialize some variables
+        # whether the customer chooses to continue loading
+        continue_loading = False
+        time_spent_on_loading = 0
```
```suggestion
```

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -811,12 +823,48 @@ def progress_patch(self, *args, **kwargs):
+        print_styled_text([(Style.ACTION, "Loading command table... Expected time around 1 minute.")])
+        progress_bar = IndeterminateProgressBar(cli_ctx=self.cli_ctx, message="Loading command table")
+        # initialize some variables
+        # whether the customer chooses to continue loading
+        continue_loading = False
+        time_spent_on_loading = 0
```
```suggestion
```

zhoxing-ms replied:
The temporary variables in a loop do not need to be defined repeatedly outside the loop

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -811,12 +823,48 @@ def progress_patch(self, *args, **kwargs):
+        # whether the customer has been prompted to choose continue loading
+        # unable to use continue_loading to check this because the customer may choose to continue loading
+        already_prompted = False
+        # whether the loading time is too long(>150s)
+        time_loading_too_long = False
```
```suggestion
```

zhoxing-ms replied:
Same as above

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -811,12 +823,48 @@ def progress_patch(self, *args, **kwargs):
+            progress_bar.update_progress()
+            time.sleep(0.1)
+            # setup how long to wait before prompting the customer to continue loading
+            if time_spent_on_loading >= 150:
```
```suggestion
            # whether the loading time is too long(>150s)
            if time_spent_on_loading >= 150:
```

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -32,12 +44,43 @@ def update(self, sample: str):
+        """Update the value of a parameter in the customized parameter value map"""
+        # remove the 'az ' part of command
+        command_text = command_text.split('az ')[-1]
+        cur_param = ''
```
Personally, I think it would be more accurate to name it as `cached_param`
```suggestion
        cached_param = ''
```

wu5bocheng replied:
Received, already modified in https://github.com/wu5bocheng/azure-cli-extensions/pull/2/commits/f92b35511e194d8a6630747c490e364c43b2039a

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -10,11 +10,23 @@
         self.cur_sample = ''
         self.cur_command = ''
+        # param_value_map is the dictionary that has `key` with command params and `value` with <sample_value>
         self.param_value_map = {}
```
Could we consider renaming it to `param_sample_value_map`?

wu5bocheng replied:
Received, already modified in https://github.com/wu5bocheng/azure-cli-extensions/commit/f92b35511e194d8a6630747c490e364c43b2039a

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/recommendation.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
[Update] optimize param values

wu5bocheng committed:
[Update] Optimize feedback function and scenario suggestion

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/scenario_suggest.py:
```diff
@@ -10,37 +10,80 @@
+        """Update the value of a parameter in the customized parameter value map"""
+        # remove the 'az ' part of command
+        command_text = command_text.split('az ')[-1]
+        param_sample_value_map = ''
```
Why is this string type variable named `xxx_map`?

zhoxing-ms replied:
How about naming it as `cached_sample_value`?

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -811,12 +823,44 @@ def progress_patch(self, *args, **kwargs):
+            time.sleep(0.1)
+            # setup how long to wait before prompting the customer to continue loading
+            # whether the loading time is too long(>150s)
+            if time_spent_on_loading >= 150:
+                time_loading_too_long = True
+            if time_loading_too_long and already_prompted == False:
```
```suggestion
            prompt_timeout_limit = 150
            if time_loading_too_long > prompt_timeout_limit and already_prompted == False:
```

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
[Update] Optimize feedback function and scenario suggestion

wu5bocheng committed:
[Update] Optimize param names

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/recommendation.py

Received, already modified.

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/recommendation.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Merge pull request #4 from wu5bocheng/telemetry_update

[Update] Update telemetry feedback of az interactive

wu5bocheng committed:
[Update] Update readme file of az interactive to describe the recommendation function change

wu5bocheng committed:
[Fix] fix bug in loading commands

wu5bocheng committed:
[Fix] fix bug in telemetry feedback

wu5bocheng committed:
[Breaking Change] Add scenario search function to az interactive

wu5bocheng committed:
[Breaking Change] Add scenario search function to az interactive

wu5bocheng committed:
[Fix] fix telemetry bug

wu5bocheng committed:
[Update] fix scenario autocomplete bugs

wu5bocheng committed:
[Update] fix warning bugs

wu5bocheng committed:
[Fix] fix scenario recommend for bool params

wu5bocheng committed:
[Update] update telemetry function name

wu5bocheng committed:
[Update] update scenario search

wu5bocheng committed:
[Update] update example output format

wu5bocheng committed:
[Update] update version

wu5bocheng committed:
[Update] Update README.md and README.rst

wu5bocheng committed:
Update src/interactive/README.md

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/recommendation.py

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/recommendation.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
Update src/interactive/README.md

wu5bocheng committed:
[Update] Update to support ctrl C

wu5bocheng committed:
[Update] Update to support ctrl C to clear screen

wu5bocheng committed:
[Delete] delete timeout in scenario search

wu5bocheng committed:
Merge pull request #7 from wu5bocheng/scenario-search

[Revolutionary] Add scenario search function to az interactive

zhoxing-ms commented on src/interactive/README.rst:
```diff
@@ -4,18 +4,21 @@ Azure CLI Interactive Shell
-* Navigation of example pane 
-* Optional layout configurations 
-* Fun Colors 
+-  Recommend Commands and Scenarios Based on Previous Commands
+-  Search for Scenarios
+-  Value Completion in Scenario Mode
+-  Interactive Tutorials
+-  Lightweight Drop Down Completions
+-  Auto Cached Suggestions
+-  Dynamic parameter completion
```
Please remove the `README.rst` and modify the `setup.py`

wu5bocheng committed:
[Update] Update to support ctrl C to clear screen

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -594,7 +616,58 @@ def scenario_repl(self, scenario):
-    # pylint: disable=too-many-statements
+    def handle_search(self, text):
+        """ parses for the scenario search """
+        # \ connect a momgodb -> connect a mongodb
```
```suggestion
        # If the user's input text is "\ connect a momgodb", we extract the keyword "connect a mongodb" from it
```

wu5bocheng committed:
[Update] Update some notes

wu5bocheng committed:
Optimize some notes

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -594,7 +616,58 @@ def scenario_repl(self, scenario):
+                if self.recommender.cur_thread.result:
+                    break
+            except (KeyboardInterrupt, ValueError):
+                # CTRL C
```
```suggestion
                # Catch CTRL + C to quit the search thread
```

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/app.py:
```diff
@@ -594,7 +616,58 @@ def scenario_repl(self, scenario):
+                print_styled_text([(Style.WARNING, "We currently can't find the scenario you need. \n"
+                                                   "You can try to change the search keywords or "
+                                                   "submit an issue to ask for the scenario you need.")])
+                # -1 means no result, since the idx+1 in feedback function, so passed -2
+                self.recommender.feedback_search(-2, keywords)
```
Can `index` be directly transmitted from `feedback_search` to `send_feedback`?

zhoxing-ms commented on src/interactive/azext_interactive/azclishell/recommendation.py:
```diff
@@ -92,12 +98,31 @@ def feedback_scenario(self, scenario_idx, scenario=None):
+            api_version = self.cur_thread.api_version
+            search_keywords = keywords.split(' ')
+            processed_exception = self.cur_thread.processed_exception
+            send_feedback(f'c{search_idx + 1}', search_keywords, processed_exception, recommendations,
+                          accepted_recommend=scenario, api_version=api_version, request_type=5)
```
Same as https://github.com/wu5bocheng/azure-cli-extensions/pull/2/files#r1163939233

wu5bocheng committed:
Update src/interactive/azext_interactive/azclishell/app.py

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>

wu5bocheng committed:
[Update] Update HISTORY.rst

wu5bocheng committed:
[Update] Update feedback logistic

wu5bocheng committed:
[Update] Update feedback function

wu5bocheng committed:
[Update] Update notes

wu5bocheng committed:
[Optimize] let uncustomized values such as -o tsv be completed automatically

zhoxing-ms commented on src/interactive/HISTORY.rst:
```diff
@@ -3,12 +3,20 @@
+* Add value completion mechanism for param value in scenarios
+* Support searching ability for scenarios
+* Optimize param completion mechanism to recommend one parameter at a time
+* Fix some small bugs
```
```suggestion
```

wu5bocheng committed:
Update src/interactive/HISTORY.rst

Co-authored-by: Xing Zhou <Zhou.Xing@microsoft.com>
